File: jas/element_test.py
from absl.testing import absltest

from element import (
    Color, Fill, Stroke, LineCap, LineJoin, Transform,
    MoveTo, LineTo, CurveTo, SmoothCurveTo, QuadTo, SmoothQuadTo, ArcTo, ClosePath,
    Line, Rect, Circle, Ellipse, Polyline, Polygon, Path, Text, Group, Layer,
)


class ElementTest(absltest.TestCase):
    """Test SVG-conforming immutable document elements."""

    def test_color_defaults(self):
        c = Color(1.0, 0.0, 0.0)
        self.assertEqual(c.a, 1.0)

    def test_color_immutable(self):
        c = Color(1.0, 0.0, 0.0)
        with self.assertRaises(AttributeError):
            c.r = 0.5

    def test_stroke_defaults(self):
        s = Stroke(Color(0, 0, 0))
        self.assertEqual(s.width, 1.0)
        self.assertEqual(s.linecap, LineCap.BUTT)
        self.assertEqual(s.linejoin, LineJoin.MITER)

    def test_transform_identity(self):
        t = Transform()
        self.assertEqual((t.a, t.b, t.c, t.d, t.e, t.f), (1, 0, 0, 1, 0, 0))

    def test_transform_translate(self):
        t = Transform.translate(10, 20)
        self.assertEqual(t.e, 10)
        self.assertEqual(t.f, 20)

    def test_transform_scale(self):
        t = Transform.scale(2, 3)
        self.assertEqual(t.a, 2)
        self.assertEqual(t.d, 3)

    def test_transform_rotate(self):
        t = Transform.rotate(90)
        self.assertAlmostEqual(t.a, 0, places=10)
        self.assertAlmostEqual(t.b, 1, places=10)

    def test_transform_scale_uniform(self):
        t = Transform.scale(3)
        self.assertEqual(t.a, 3)
        self.assertEqual(t.d, 3)

    def test_line_bounds(self):
        ln = Line(x1=0, y1=0, x2=10, y2=20)
        self.assertEqual(ln.bounds(), (0, 0, 10, 20))

    def test_line_reversed(self):
        ln = Line(x1=10, y1=20, x2=0, y2=0)
        self.assertEqual(ln.bounds(), (0, 0, 10, 20))

    def test_line_immutable(self):
        ln = Line(x1=0, y1=0, x2=10, y2=10)
        with self.assertRaises(AttributeError):
            ln.x1 = 5

    def test_rect_bounds(self):
        r = Rect(x=5, y=10, width=100, height=50)
        self.assertEqual(r.bounds(), (5, 10, 100, 50))

    def test_rect_rounded(self):
        r = Rect(x=0, y=0, width=10, height=10, rx=2, ry=2)
        self.assertEqual(r.rx, 2)
        self.assertEqual(r.ry, 2)

    def test_rect_immutable(self):
        r = Rect(x=0, y=0, width=10, height=10)
        with self.assertRaises(AttributeError):
            r.x = 5

    def test_circle_bounds(self):
        c = Circle(cx=50, cy=50, r=25)
        self.assertEqual(c.bounds(), (25, 25, 50, 50))

    def test_circle_with_fill_and_stroke(self):
        c = Circle(cx=50, cy=50, r=25,
                   fill=Fill(Color(0, 1, 0)),
                   stroke=Stroke(Color(0, 0, 0), width=3.0))
        self.assertEqual(c.fill.color.g, 1.0)
        self.assertEqual(c.stroke.width, 3.0)

    def test_ellipse_bounds(self):
        e = Ellipse(cx=50, cy=50, rx=25, ry=15)
        self.assertEqual(e.bounds(), (25, 35, 50, 30))

    def test_ellipse_with_fill_and_stroke(self):
        e = Ellipse(cx=50, cy=50, rx=25, ry=15,
                    fill=Fill(Color(0, 0, 1)),
                    stroke=Stroke(Color(1, 1, 1), linecap=LineCap.SQUARE))
        self.assertEqual(e.fill.color.b, 1.0)
        self.assertEqual(e.stroke.linecap, LineCap.SQUARE)

    def test_polyline_bounds(self):
        pl = Polyline(points=((0, 0), (10, 5), (20, 0)))
        self.assertEqual(pl.bounds(), (0, 0, 20, 5))

    def test_empty_polyline(self):
        pl = Polyline(points=())
        self.assertEqual(pl.bounds(), (0, 0, 0, 0))

    def test_polygon_bounds(self):
        pg = Polygon(points=((0, 0), (10, 0), (5, 10)))
        self.assertEqual(pg.bounds(), (0, 0, 10, 10))

    def test_empty_polygon(self):
        pg = Polygon(points=())
        self.assertEqual(pg.bounds(), (0, 0, 0, 0))

    def test_path_bounds(self):
        p = Path(d=(MoveTo(0, 0), LineTo(10, 20), LineTo(5, 15), ClosePath()))
        self.assertEqual(p.bounds(), (0, 0, 10, 20))

    def test_path_cubic_bezier(self):
        p = Path(d=(MoveTo(0, 0), CurveTo(5, 10, 15, 10, 20, 0)))
        self.assertEqual(p.bounds(), (0, 0, 20, 0))

    def test_path_smooth_curve_to(self):
        p = Path(d=(MoveTo(0, 0), CurveTo(1, 2, 3, 4, 5, 6), SmoothCurveTo(8, 9, 10, 12)))
        self.assertEqual(p.bounds(), (0, 0, 10, 12))

    def test_path_quad_to(self):
        p = Path(d=(MoveTo(0, 0), QuadTo(5, 10, 10, 0)))
        self.assertEqual(p.bounds(), (0, 0, 10, 0))

    def test_path_smooth_quad_to(self):
        p = Path(d=(MoveTo(0, 0), QuadTo(5, 10, 10, 0), SmoothQuadTo(20, 5)))
        self.assertEqual(p.bounds(), (0, 0, 20, 5))

    def test_path_arc_to(self):
        p = Path(d=(MoveTo(0, 0), ArcTo(rx=25, ry=25, x_rotation=0, large_arc=True, sweep=False, x=50, y=0)))
        self.assertEqual(p.bounds(), (0, 0, 50, 0))

    def test_path_empty(self):
        p = Path(d=())
        self.assertEqual(p.bounds(), (0, 0, 0, 0))

    def test_path_with_fill_and_stroke(self):
        fill = Fill(Color(1, 0, 0))
        stroke = Stroke(Color(0, 0, 0), width=2.0, linecap=LineCap.ROUND)
        p = Path(d=(MoveTo(0, 0), LineTo(10, 10), ClosePath()), fill=fill, stroke=stroke)
        self.assertEqual(p.fill.color.r, 1.0)
        self.assertEqual(p.stroke.width, 2.0)
        self.assertEqual(p.stroke.linecap, LineCap.ROUND)

    def test_text_bounds(self):
        t = Text(x=10, y=30, content="Hello")
        x, y, w, h = t.bounds()
        self.assertEqual(x, 10)
        self.assertEqual(y, 14)  # y - font_size
        self.assertGreater(w, 0)
        self.assertEqual(h, 16)

    def test_text_attributes(self):
        t = Text(x=0, y=0, content="Hi", font_family="monospace", font_size=24.0)
        self.assertEqual(t.font_family, "monospace")
        self.assertEqual(t.font_size, 24.0)

    def test_group_bounds(self):
        r = Rect(x=0, y=0, width=10, height=10)
        e = Ellipse(cx=100, cy=100, rx=5, ry=5)
        g = Group(children=(r, e))
        self.assertEqual(g.bounds(), (0, 0, 105, 105))

    def test_group_empty(self):
        g = Group(children=())
        self.assertEqual(g.bounds(), (0, 0, 0, 0))

    def test_nested_group(self):
        inner = Group(children=(Rect(x=10, y=10, width=5, height=5),))
        outer = Group(children=(Rect(x=0, y=0, width=1, height=1), inner))
        self.assertEqual(outer.bounds(), (0, 0, 15, 15))

    def test_group_with_transform(self):
        g = Group(children=(Rect(x=0, y=0, width=10, height=10),),
                  transform=Transform.translate(100, 200))
        self.assertIsNotNone(g.transform)
        self.assertEqual(g.transform.e, 100)

    def test_element_opacity(self):
        r = Rect(x=0, y=0, width=10, height=10, opacity=0.5)
        self.assertEqual(r.opacity, 0.5)

    def test_group_all_element_types(self):
        children = (
            Line(x1=0, y1=0, x2=10, y2=10),
            Rect(x=0, y=0, width=20, height=20),
            Circle(cx=50, cy=50, r=10),
            Ellipse(cx=50, cy=50, rx=10, ry=5),
            Polyline(points=((0, 0), (10, 10))),
            Polygon(points=((0, 0), (10, 0), (5, 10))),
            Path(d=(MoveTo(0, 0), LineTo(10, 10))),
            Text(x=0, y=20, content="test"),
        )
        g = Group(children=children)
        x, y, w, h = g.bounds()
        self.assertEqual(x, 0)
        self.assertEqual(y, 0)
        self.assertGreater(w, 0)
        self.assertGreater(h, 0)

    def test_deeply_nested_groups(self):
        inner = Group(children=(Rect(x=10, y=10, width=5, height=5),))
        mid = Group(children=(Rect(x=0, y=0, width=1, height=1), inner))
        outer = Group(children=(Rect(x=20, y=20, width=3, height=3), mid))
        x, y, w, h = outer.bounds()
        self.assertEqual(x, 0)
        self.assertEqual(y, 0)
        self.assertEqual(w, 23)
        self.assertEqual(h, 23)

    def test_layer_default_name(self):
        layer = Layer(children=(Rect(x=0, y=0, width=10, height=10),))
        self.assertEqual(layer.name, "Layer")

    def test_layer_custom_name(self):
        layer = Layer(children=(Rect(x=0, y=0, width=10, height=10),), name="Background")
        self.assertEqual(layer.name, "Background")

    def test_layer_bounds(self):
        layer = Layer(children=(
            Rect(x=0, y=0, width=10, height=10),
            Circle(cx=50, cy=50, r=5),
        ), name="Shapes")
        self.assertEqual(layer.bounds(), (0, 0, 55, 55))

    def test_layer_empty(self):
        layer = Layer(children=(), name="Empty")
        self.assertEqual(layer.bounds(), (0, 0, 0, 0))

    def test_layer_is_group(self):
        layer = Layer(children=(), name="Test")
        self.assertIsInstance(layer, Group)


if __name__ == "__main__":
    absltest.main()

File: jas_ocaml/test/element_test.ml
let () =
  let open Jas.Element in

  (* Test color construction *)
  let c = make_color 1.0 0.0 0.0 in
  assert (c.r = 1.0 && c.a = 1.0);
  let c2 = make_color ~a:0.5 0.0 1.0 0.0 in
  assert (c2.a = 0.5);

  (* Test stroke defaults *)
  let s = make_stroke (make_color 0.0 0.0 0.0) in
  assert (s.stroke_width = 1.0);
  assert (s.stroke_linecap = Butt);
  assert (s.stroke_linejoin = Miter);

  (* Test transform helpers *)
  let t = identity_transform in
  assert (t.a = 1.0 && t.d = 1.0 && t.e = 0.0);
  let t2 = make_translate 10.0 20.0 in
  assert (t2.e = 10.0 && t2.f = 20.0);
  let t3 = make_scale 2.0 3.0 in
  assert (t3.a = 2.0 && t3.d = 3.0);

  (* Test line bounds *)
  let ln = make_line 0.0 0.0 10.0 20.0 in
  let (x, y, w, h) = bounds ln in
  assert (x = 0.0 && y = 0.0 && w = 10.0 && h = 20.0);

  (* Test rect bounds *)
  let r = make_rect 5.0 10.0 100.0 50.0 in
  let (x, y, w, h) = bounds r in
  assert (x = 5.0 && y = 10.0 && w = 100.0 && h = 50.0);

  (* Test rect rounded corners *)
  let rr = make_rect ~rx:2.0 ~ry:2.0 0.0 0.0 10.0 10.0 in
  (match rr with Rect { rx; ry; _ } -> assert (rx = 2.0 && ry = 2.0) | _ -> assert false);

  (* Test circle bounds *)
  let ci = make_circle 50.0 50.0 25.0 in
  let (x, y, w, h) = bounds ci in
  assert (x = 25.0 && y = 25.0 && w = 50.0 && h = 50.0);

  (* Test ellipse bounds *)
  let el = make_ellipse 50.0 50.0 25.0 15.0 in
  let (x, y, w, h) = bounds el in
  assert (x = 25.0 && y = 35.0 && w = 50.0 && h = 30.0);

  (* Test polyline bounds *)
  let pl = make_polyline [(0.0, 0.0); (10.0, 5.0); (20.0, 0.0)] in
  let (x, y, w, h) = bounds pl in
  assert (x = 0.0 && y = 0.0 && w = 20.0 && h = 5.0);

  (* Test polygon bounds *)
  let pg = make_polygon [(0.0, 0.0); (10.0, 0.0); (5.0, 10.0)] in
  let (x, y, w, h) = bounds pg in
  assert (x = 0.0 && y = 0.0 && w = 10.0 && h = 10.0);

  (* Test path bounds with SVG commands *)
  let p = make_path [MoveTo (0.0, 0.0); LineTo (10.0, 20.0); LineTo (5.0, 15.0); ClosePath] in
  let (x, y, w, h) = bounds p in
  assert (x = 0.0 && y = 0.0 && w = 10.0 && h = 20.0);

  (* Test path with cubic bezier *)
  let p2 = make_path [MoveTo (0.0, 0.0); CurveTo (5.0, 10.0, 15.0, 10.0, 20.0, 0.0)] in
  let (x, _, w, _) = bounds p2 in
  assert (x = 0.0 && w = 20.0);

  (* Test empty path *)
  let ep = make_path [] in
  let (x, y, w, h) = bounds ep in
  assert (x = 0.0 && y = 0.0 && w = 0.0 && h = 0.0);

  (* Test path with fill and stroke *)
  let styled = make_path
    ~fill:(Some (make_fill (make_color 1.0 0.0 0.0)))
    ~stroke:(Some (make_stroke ~width:2.0 ~linecap:Round_cap (make_color 0.0 0.0 0.0)))
    [MoveTo (0.0, 0.0); LineTo (10.0, 10.0); ClosePath] in
  (match styled with
   | Path { fill; stroke; _ } ->
     (match fill with Some f -> assert (f.fill_color.r = 1.0) | None -> assert false);
     (match stroke with Some s -> assert (s.stroke_width = 2.0 && s.stroke_linecap = Round_cap) | None -> assert false)
   | _ -> assert false);

  (* Test text bounds *)
  let txt = make_text 10.0 30.0 "Hello" in
  let (x, y, w, h) = bounds txt in
  assert (x = 10.0 && y = 14.0 && w > 0.0 && h = 16.0);

  (* Test text attributes *)
  let txt2 = make_text ~font_family:"monospace" ~font_size:24.0 0.0 0.0 "Hi" in
  (match txt2 with Text { font_family; font_size; _ } -> assert (font_family = "monospace" && font_size = 24.0) | _ -> assert false);

  (* Test group bounds *)
  let g = make_group [r; el] in
  let (x, y, w, h) = bounds g in
  assert (x = 5.0 && y = 10.0 && w = 100.0 && h = 55.0);

  (* Test empty group *)
  let eg = make_group [] in
  let (x, y, w, h) = bounds eg in
  assert (x = 0.0 && y = 0.0 && w = 0.0 && h = 0.0);

  (* Test nested group *)
  let inner = make_group [make_rect 10.0 10.0 5.0 5.0] in
  let outer = make_group [make_rect 0.0 0.0 1.0 1.0; inner] in
  let (x, y, w, h) = bounds outer in
  assert (x = 0.0 && y = 0.0 && w = 15.0 && h = 15.0);

  (* Test group with transform *)
  let gt = make_group ~transform:(Some (make_translate 100.0 200.0)) [make_rect 0.0 0.0 10.0 10.0] in
  (match gt with Group { transform; _ } ->
    (match transform with Some t -> assert (t.e = 100.0) | None -> assert false)
   | _ -> assert false);

  (* Test element opacity *)
  let ro = make_rect ~opacity:0.5 0.0 0.0 10.0 10.0 in
  (match ro with Rect { opacity; _ } -> assert (opacity = 0.5) | _ -> assert false);

  (* Test path with SmoothCurveTo *)
  let psc = make_path [MoveTo (0.0, 0.0); CurveTo (1.0, 2.0, 3.0, 4.0, 5.0, 6.0); SmoothCurveTo (8.0, 9.0, 10.0, 12.0)] in
  let (x, _, w, h) = bounds psc in
  assert (x = 0.0 && w = 10.0 && h = 12.0);

  (* Test path with QuadTo *)
  let pq = make_path [MoveTo (0.0, 0.0); QuadTo (5.0, 10.0, 10.0, 0.0)] in
  let (x, _, w, _) = bounds pq in
  assert (x = 0.0 && w = 10.0);

  (* Test path with SmoothQuadTo *)
  let psq = make_path [MoveTo (0.0, 0.0); QuadTo (5.0, 10.0, 10.0, 0.0); SmoothQuadTo (20.0, 5.0)] in
  let (x, _, w, _) = bounds psq in
  assert (x = 0.0 && w = 20.0);

  (* Test path with ArcTo *)
  let pa = make_path [MoveTo (0.0, 0.0); ArcTo (25.0, 25.0, 0.0, true, false, 50.0, 0.0)] in
  let (x, _, w, _) = bounds pa in
  assert (x = 0.0 && w = 50.0);

  (* Test empty polyline *)
  let epl = make_polyline [] in
  let (x, y, w, h) = bounds epl in
  assert (x = 0.0 && y = 0.0 && w = 0.0 && h = 0.0);

  (* Test empty polygon *)
  let epg = make_polygon [] in
  let (x, y, w, h) = bounds epg in
  assert (x = 0.0 && y = 0.0 && w = 0.0 && h = 0.0);

  (* Test reversed line coordinates *)
  let rl = make_line 10.0 20.0 0.0 0.0 in
  let (x, y, w, h) = bounds rl in
  assert (x = 0.0 && y = 0.0 && w = 10.0 && h = 20.0);

  (* Test circle with fill and stroke *)
  let cfs = make_circle
    ~fill:(Some (make_fill (make_color 0.0 1.0 0.0)))
    ~stroke:(Some (make_stroke ~width:3.0 (make_color 0.0 0.0 0.0)))
    50.0 50.0 25.0 in
  (match cfs with
   | Circle { fill; stroke; _ } ->
     (match fill with Some f -> assert (f.fill_color.g = 1.0) | None -> assert false);
     (match stroke with Some s -> assert (s.stroke_width = 3.0) | None -> assert false)
   | _ -> assert false);

  (* Test ellipse with fill and stroke *)
  let efs = make_ellipse
    ~fill:(Some (make_fill (make_color 0.0 0.0 1.0)))
    ~stroke:(Some (make_stroke ~linecap:Square (make_color 1.0 1.0 1.0)))
    50.0 50.0 25.0 15.0 in
  (match efs with
   | Ellipse { fill; stroke; _ } ->
     (match fill with Some f -> assert (f.fill_color.b = 1.0) | None -> assert false);
     (match stroke with Some s -> assert (s.stroke_linecap = Square) | None -> assert false)
   | _ -> assert false);

  (* Test transform rotate *)
  let tr = make_rotate 90.0 in
  assert (abs_float tr.a < 1e-10);
  assert (abs_float (tr.b -. 1.0) < 1e-10);

  (* Test group with all element types *)
  let all_types = make_group [
    make_line 0.0 0.0 10.0 10.0;
    make_rect 0.0 0.0 20.0 20.0;
    make_circle 50.0 50.0 10.0;
    make_ellipse 50.0 50.0 10.0 5.0;
    make_polyline [(0.0, 0.0); (10.0, 10.0)];
    make_polygon [(0.0, 0.0); (10.0, 0.0); (5.0, 10.0)];
    make_path [MoveTo (0.0, 0.0); LineTo (10.0, 10.0)];
    make_text 0.0 20.0 "test";
  ] in
  let (x, _, w, h) = bounds all_types in
  assert (x = 0.0 && w > 0.0 && h > 0.0);

  (* Test deeply nested groups (3 levels) *)
  let deep_inner = make_group [make_rect 10.0 10.0 5.0 5.0] in
  let deep_mid = make_group [make_rect 0.0 0.0 1.0 1.0; deep_inner] in
  let deep_outer = make_group [make_rect 20.0 20.0 3.0 3.0; deep_mid] in
  let (x, y, w, h) = bounds deep_outer in
  assert (x = 0.0 && y = 0.0 && w = 23.0 && h = 23.0);

  (* Test layer default name *)
  let layer = make_layer [make_rect 0.0 0.0 10.0 10.0] in
  (match layer with Layer { name; _ } -> assert (name = "Layer") | _ -> assert false);

  (* Test layer custom name *)
  let layer2 = make_layer ~name:"Background" [make_rect 0.0 0.0 10.0 10.0] in
  (match layer2 with Layer { name; _ } -> assert (name = "Background") | _ -> assert false);

  (* Test layer bounds *)
  let layer3 = make_layer ~name:"Shapes" [make_rect 0.0 0.0 10.0 10.0; make_circle 50.0 50.0 5.0] in
  let (x, y, w, h) = bounds layer3 in
  assert (x = 0.0 && y = 0.0 && w = 55.0 && h = 55.0);

  (* Test empty layer *)
  let layer4 = make_layer ~name:"Empty" [] in
  let (x, y, w, h) = bounds layer4 in
  assert (x = 0.0 && y = 0.0 && w = 0.0 && h = 0.0);

  Printf.printf "All element tests passed.\n"
